scripts/src/java/org/oppia/android/scripts/coverage/RunCoverage.kt

```diff
@@ -0,0 +1,154 @@
+   * prints a message indicating no coverage analysis is performed. Otherwise, initializes
+   * a Bazel client, finds potential test file paths, retrieves Bazel targets, and initiates
+   * coverage analysis for each test target found.
+   */
+  fun execute(): MutableList<List<String>> {
+    var coverageDataList = mutableListOf<List<String>>()
+    val testFileExemptionList = loadTestFileExemptionsProto(testFileExemptionTextProto)
+      .getExemptedFilePathList()
+
+    val isExempted = testFileExemptionList.contains(filePath)
+    if (isExempted) {
+      println("This file is exempted from having a test file. Hence No coverage!")
+      return mutableListOf()
+    }
+
+    val testFilePaths = findTestFile(repoRoot, filePath)
+    val testTargets = bazelClient.retrieveBazelTargets(testFilePaths)
+
+    for (testTarget in testTargets) {
+      val coverageData = RunCoverageForTestTarget(
+        rootDirectory,
+        testTarget.substringBeforeLast(".kt"),
+        commandExecutor,
+        scriptBgDispatcher
+      ).runCoverage()!!
+      coverageDataList.add(coverageData)
+    }
+    println("Coverage Data List: $coverageDataList")
+    return coverageDataList
+  }
+
+  /**
+   * Finds potential test file paths corresponding to a given source file path within a repository.
+   *
+   * @param repoRoot the root directory of the repository
+   * @param filePath The file path of the source file for which the test files are to be found.
+   * @return A list of potential test file paths that exist in the repository.
+   */
+  fun findTestFile(repoRoot: String, filePath: String): List<String> {
+    val file = File(filePath)
+    val parts = file.parent.split(File.separator)
+    val testFiles = mutableListOf<String>()
+
+    if (parts.isNotEmpty() && parts[0] == "scripts") {
+      val testFilePath = filePath.replace("/java/", "/javatests/").replace(".kt", "Test.kt")
+      if (File(repoRoot, testFilePath).exists()) {
+        testFiles.add(testFilePath)
+      }
+    } else if (parts.isNotEmpty() && parts[0] == "app") {
+      val sharedTestFilePath = filePath.replace("/main/", "/sharedTest/").replace(".kt", "Test.kt")
+      val testFilePath = filePath.replace("/main/", "/test/").replace(".kt", "Test.kt")
+      val localTestFilePath = filePath.replace("/main/", "/test/").replace(".kt", "LocalTest.kt")
+
+      if (File(repoRoot, sharedTestFilePath).exists()) {
+        testFiles.add(sharedTestFilePath)
+      }
+      if (File(repoRoot, testFilePath).exists()) {
+        testFiles.add(testFilePath)
+      }
+      if (File(repoRoot, localTestFilePath).exists()) {
+        testFiles.add(localTestFilePath)
+      }
+    } else {
+      val defaultTestFilePath = filePath.replace("/main/", "/test/").replace(".kt", "Test.kt")
+      if (File(repoRoot, defaultTestFilePath).exists()) {
+        testFiles.add(defaultTestFilePath)
+      }
+    }
+    return testFiles
+  }
```

Comment:
```suggestion
    val possibleTestFilePaths = when {
      filePath.startsWith("scripts/") ->
        listOf(filePath.replace("/java/", "/javatests/").replace(".kt", "Test.kt"))
      filePath.startsWith("app/") -> {
        listOf(
          filePath.replace("/main/", "/sharedTest/").replace(".kt", "Test.kt"),
          filePath.replace("/main/", "/test/").replace(".kt", "Test.kt"),
          filePath.replace("/main/", "/test/").replace(".kt", "LocalTest.kt")
        )
      }
      else -> listOf(filePath.replace("/main/", "/test/").replace(".kt", "Test.kt"))
    }
    return possibleTestFilePaths.map { File(repoRoot, it) }.filter(File::exists)
  }
```

Some Kotlin patterns can simplify this somewhat by removing duplication and avoiding using a mutable structure, and possibly make it a bit easier to follow as a reader.

Reply:
Thanks @BenHenning, that was super helpful! Updated with the provided pattern with the addition to just convert and return them as relative paths.